src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -851,6 +851,22 @@ public INamedTypeSymbol GetTypeByMetadataName(string fullyQualifiedMetadataName)
         /// </summary>
         public abstract ImmutableArray<Diagnostic> GetDiagnostics(CancellationToken cancellationToken = default(CancellationToken));
 
+        internal void EnsureCompilationCompleted()
+        {
+            Debug.Assert(this.EventQueue != null);
+
+            lock (this.EventQueue)
```

Comment:
This is needed as multiple threads might call compilation.GetDiagnostics, which in turn invokes this method when there are no syntax trees.